### PR TITLE
cntk: fix build and add comments

### PR DESCRIPTION
**Pull request by glennpj**

### CODE REVIEW

**Review by alalazo — DISMISSED** [review]
LGTM, but wanted to double check that the intent is to deprecate _all versions_

**Comment by glennpj** [comment]
> LGTM, but wanted to double check that the intent is to deprecate _all versions_

Perhaps the comments that I put in the file are sufficient rather than deprecating?

**Comment by alalazo** [comment]
> Perhaps the comments that I put in the file are sufficient rather than deprecating?

I am asking since currently one needs to install with `--deprecated` to have Spack use a deprecated version without being prompted at fetching time:
```console
$ spack install openssl@1.1.1l
==> Warning: using "openssl@1.1.1l" which is a deprecated version
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/berkeley-db-18.1.40-jjcdgicmjttpteerizjgcko7ivhxp7yh
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/libiconv-1.16-klh4setxxl7qui3sr56bjraw3qskydqu
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/pkgconf-1.8.0-k7oj7j7cv6qkren5nu2zpaknr3at4d2z
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/zlib-1.2.11-h7tvzacqculnkfs44qyfaizurnqsfvdr
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/diffutils-3.8-b6dr6dxubin6lktaamjxqwdvwii7ngrt
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/ncurses-6.2-vc37m24s7t6af7vwhhlvhd22lmbchqib
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/bzip2-1.0.8-iwjq7ca2gxdipsqojnb2h5xn3lxz2orb
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/readline-8.1-autymkx77iesq7hbpwiwg5il2j52zrer
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/gdbm-1.19-spq3vzbvhjcnu7admmbrijoami454pgl
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-11.2.0/perl-5.34.0-esrjoliefurio2rnhkuzivvhyawkcdzq
==> Installing openssl-1.1.1l-roiqch4bpn5kopz6zcnde5awjb6r62c3
==> No binary for openssl-1.1.1l-roiqch4bpn5kopz6zcnde5awjb6r62c3 found: installing from source
==> Warning: openssl@1.1.1l is deprecated and may be removed in a future Spack release.
==>   Fetch anyway? [y/N] 
```
Whether or not that is what we want I think it's a decision of the maintainers of the package and the community. I never used `cntk` so I have no opinion here. Maybe @adamjstewart has some thoughts?

File: var/spack/repos/builtin/packages/cntk/package.py [hunk 1]
```diff
@@ -49,6 +54,26 @@ class Cntk(Package):
     # Patch to change behaviour of lock file - https://github.com/Microsoft/CNTK/issues/62
     patch('lock-file.patch')
 
+    # It seems that cntk, at least version 2.0, can not be built with GCC
+    # beyond 4.8.5.
+    conflicts('%gcc@5:')
+
+    def patch(self):
+        if 'protobuf+shared' in self.spec:
+            library_suffix = 'so'
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cntk/package.py]: Is this also true on macOS?

File: var/spack/repos/builtin/packages/cntk/package.py [hunk 2]
```diff
@@ -49,6 +54,26 @@ class Cntk(Package):
...
+        else:
+            library_suffix = 'a'
+
+        protobuf_libdir = os.path.basename(self.spec['protobuf'].libs.directories[0])
+        protobuf_ld_flags = self.spec['protobuf'].libs.ld_flags
+
+        filter_file(r'(protobuf_check=)lib/(libprotobuf\.)a',
+                    r'\1{0}/\2{1}'.format(protobuf_libdir, library_suffix),
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/cntk/package.py]: You should be able to use `spec['protobuf'].libs[0]` if you just want the full path to the library file

**Review by adamjstewart — COMMENTED** [review]
Many of the packages in Spack are no longer actively developed. We don't necessarily need to deprecate and remove this package, there might be people who still use it for reproducibility of old ML experiments.

File: var/spack/repos/builtin/packages/cntk/package.py (hunk 2 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/cntk/package.py]: Due to the way the check is in the configure script, it needs to be split. But your comment gave me an idea of how to do this better and eliminate the need for the suffix variable.

File: var/spack/repos/builtin/packages/cntk/package.py (hunk 1 above)
Comment (glennpj) [reply var/spack/repos/builtin/packages/cntk/package.py]: This will not be necessary at all with other changes being made.

**Comment by glennpj** [comment]
> Many of the packages in Spack are no longer actively developed. We don't necessarily need to deprecate and remove this package, there might be people who still use it for reproducibility of old ML experiments.

Yep, I went to the spack docs to reread the deprecation section. I will back that out.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**